zerver/forms.py

```diff
@@ -383,6 +421,57 @@ def __init__(
 
     @override
     def clean(self) -> None:
+        super().clean()
+        if not self.data.get("captcha"):
+            self.add_error("captcha", _("Validation failed, please try again."))
+
+    def clean_captcha(self) -> str:
+        payload = self.data.get("captcha", "")
+        if not settings.USING_CAPTCHA or not settings.ALTCHA_HMAC_KEY:  # nocoverage
+            raise forms.ValidationError(_("Challenges are not enabled."))
+
+        try:
+            ok, err = verify_solution(payload, settings.ALTCHA_HMAC_KEY, check_expires=True)
+            if not ok:
+                logging.warning("Invalid altcha solution: %s", err)
+                raise forms.ValidationError(_("Validation failed, please try again."))
+        except forms.ValidationError:
+            raise
+        except Exception as e:
+            logging.exception(e)
+            raise forms.ValidationError(_("Validation failed, please try again."))
+
+        payload = orjson.loads(base64.b64decode(payload))
+        challenge = payload["challenge"]
+        session_challenges = [e[0] for e in self.request.session.get("altcha_challenges", [])]
+        if challenge not in session_challenges:
+            logging.warning("Expired or replayed altcha solution")
+            raise forms.ValidationError(_("Validation failed, please try again."))
+
+        # Remove the successful solve from the session, to prevent replay
+        self.request.session["altcha_challenges"] = [
+            e for e in self.request.session.get("altcha_challenges", []) if e[0] != challenge
+        ]
+
+        return payload
+
+
+class CaptchaDemoRegistrationForm(DemoRegistrationForm):
```

Comment:
This is just a copy and paste of the captcha form logic that's currently being used to create new Zulip organizations at the moment.